NRPT-260: Adding/Editing Records for a Mine

dhlevi opened the pull request:
Added screens and functionality for adding/editing records within the context of a mine. Includes a merge from existing branch from @mtCarto for the bulk of the Add code.

Note: From the AC for NRPT-197, there is a request to auto-generate and auto-publish a record when it is added to a collection. This is not in place due to a number of technical reasons which should likely be addressed in a future ticket as the changes required go beyond the scope of simply adding/editing records for a mine.

See tickets [NRPT-260](https://bcmines.atlassian.net/browse/NRPT-260) and [NRPT-197](https://bcmines.atlassian.net/browse/NRPT-197)

dhlevi commented:
Whoops, forgot to squash before PRing. Apologies about the lengthy commit history!

KitArmstrong commented on angular/projects/admin-nrpti/src/app/mines/mines-records-add-edit/mines-records-add-edit.component.ts:
```diff
@@ -0,0 +1,349 @@
+    record['recordType'] = this.myForm.get('recordType').value;
+    this.myForm.get('recordAgency').dirty && (record['issuingAgency'] = this.myForm.get('recordAgency').value);
+
+    // lookup appropriate schemaName from from type value
```
"from from"

dhlevi replied:
Fixed

KitArmstrong commented on angular/projects/admin-nrpti/src/app/services/factory.service.ts:
```diff
@@ -350,6 +350,139 @@ export class FactoryService {
+  public editMineRecord(record: any): Observable<object> {
+    const outboundObject = {};
+
+    // theres a bunch missing?
```
Are there?

dhlevi replied:
Well, there was! That comment doesn't make as much sense anymore.

However, does bring up a note that some of the types in the list do not have BCMI flavours and therefore will not work. Those types should either have flavours added, or be removed from the list if they are not needed for BCMI.

KitArmstrong commented on angular/projects/admin-nrpti/src/app/services/factory.service.ts:
```diff
@@ -350,6 +350,139 @@ export class FactoryService {
+        break;
+    }
+
+    console.log(outboundObject);
```
Log

dhlevi replied:
Removed

KitArmstrong commented on angular/projects/admin-nrpti/src/app/services/factory.service.ts:
```diff
@@ -350,6 +350,139 @@ export class FactoryService {
+   * @returns {Promise<any>}
+   * @memberof FactoryService
+   */
+  // todo verify need model specified
```
Is this still a todo?

dhlevi replied:
Nope, fixed

KitArmstrong commented on api/src/controllers/post/dam-safety-inspection.js:
```diff
@@ -89,7 +89,7 @@ exports.createMaster = function (args, res, next, incomingObj, flavourIds) {
   // set data
   incomingObj.recordName && (damSafetyInspection.recordName = incomingObj.recordName);
-  damSafetyInspection.recordType = 'DamSafetyInspection';
+  damSafetyInspection.recordType = 'Dam Safety Inspection';
```
Nice catch, I noticed this during the last demo.

dhlevi commented:
> * I think we need a way to add a permit amendment.  Whether it is a new dropdown that appears if permit is select or if permit amendment appears as a type and just handles it all in the back, I think the option needs to be there.

In progress. Looking into the best way to implement in the add/edit UI now.

> * The Add Record screen has the title "Edit Record."

Fixed!

> * Some record types throw an error when saving.  `ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'subscribe' of null TypeError: Cannot read property 'subscribe' of null at MinesRecordsAddEditComponent.<anonymous> (mines-records-add-edit.component.ts:309)`

Couldn't reproduce, but I made a few other changes so may have been inadvertently fixed? Likely related to types without BCMI flavours, which have been removed for now until those types are updated.

> * Sometimes get a generic error when creating that just says "Failed to create record" with no indication why. Name = "Testing", Date = "2020-08-05",   Type = "Agreement", Agency = "EAO"

Couldn't reproduce, but I made a few other changes so may have been inadvertently fixed?

> * Doesn't appear to be a way to add documents to a record when creating it

Fixed!

> * Once you finish editing it directs you to the details view. The only way to get back to the records list is the use the breadcrumbs at the top.  Should there be some other action button?

Yeah, could be a good idea. Looks like that's pretty consistent in all of our Details screens. I'll leave it as-is for now to be consistent, but it would be a good idea to get a recommendation from QA and make a ticket if appropriate.

dhlevi commented:
> This code looks very similar to #538. Maybe in the future we will have time to consolidate this code.

It's probably more or less identical. Good idea to flag it as technical debt to merge into something shared in the future.

danieltruong commented:
> > This code looks very similar to #538. Maybe in the future we will have time to consolidate this code.
> 
> It's probably more or less identical. Good idea to flag it as technical debt to merge into something shared in the future.

I'll make a ticket
